run/jenkins.py

```diff
@@ -0,0 +1,282 @@
+
+    PROD_HOST = 'https://wptdashboard.appspot.com'
+    GS_RESULTS_BUCKET = 'wptd'
+    LOCAL_LOG_FILEPATH = '%s/wptd-testrun.log' % args['output_path']
+    LOCAL_REPORT_FILEPATH = "%s/wptd-%s-%s-report.log" % (
+        args['output_path'], args['SHA'], platform_id
+    )
+    SUMMARY_PATH = '%s/%s-summary.json.gz' % (args['SHA'], platform_id)
+    LOCAL_SUMMARY_GZ_FILEPATH = "%s/%s" % (args['output_path'], SUMMARY_PATH)
+    GS_RESULTS_FILEPATH_BASE = "%s/%s/%s" % (
+        args['output_path'], args['SHA'], platform_id
+    )
+    GS_HTTP_RESULTS_URL = 'https://storage.googleapis.com/%s/%s' % (
+        GS_RESULTS_BUCKET, SUMMARY_PATH
+    )
+
+    SUMMARY_FILENAME = '%s-%s-summary.json.gz' % (args['SHA'], platform_id)
+    SUMMARY_HTTP_URL = 'https://storage.googleapis.com/%s/%s' % (
+        GS_RESULTS_BUCKET, SUMMARY_FILENAME
+    )
+
+    if platform.get('sauce'):
+        if args['sauce_from_metadata']:
+            print 'Fetching Sauce creds from GCE metadata...'
+
+            for key in ('sauce_user', 'sauce_key'):
+                url = '%s/project/attributes/%s' % (METADATA_URL, key)
+                res = requests.get(url, headers={'Metadata-Flavor': 'Google'})
+                args[key] = res.text
+                assert args[key], 'Metadata key %s is empty.' % key
+
+        assert args['sauce_key'], 'SAUCE_KEY env var required'
+        assert args['sauce_user'], 'SAUCE_USER env var required'
+        SAUCE_TUNNEL_ID = '%s_%s' % (platform_id, int(time.time()))
+
+    assert len(args['SHA']) == 10, 'SHA must a WPT SHA[:10]'
+
+    # Hack because Sauce expects a different name
+    # Maybe just change it in browsers.json?
+    if platform['browser_name'] == 'edge':
+        sauce_browser_name = 'MicrosoftEdge'
+    else:
+        sauce_browser_name = platform['browser_name']
+    product = 'sauce:%s:%s' % (sauce_browser_name, platform['browser_version'])
+
+    patch_wpt(args['wptd_path'], args['wpt_path'], platform)
+
+    if platform.get('sauce'):
+        command = [
+            './wpt', 'run', product,
+            '--sauce-platform=%s' % platform['os_name'],
+            '--sauce-key=%s' % args['sauce_key'],
+            '--sauce-user=%s' % args['sauce_user'],
+            '--sauce-tunnel-id=%s' % SAUCE_TUNNEL_ID,
+            '--no-restart-on-unexpected',
+            '--processes=2',
+            '--run-by-dir=3',
+            '--log-mach=-',
+            '--log-wptreport=%s' % LOCAL_REPORT_FILEPATH,
+            '--install-fonts'
+        ]
+        if args['run_path']:
+            command.insert(3, args['run_path'])
+    else:
+        command = [
+            'xvfb-run', '--auto-servernum',
+            './wpt', 'run',
+            platform['browser_name'],
+            '--install-fonts',
+            '--install-browser',
+            '--yes',
+            '--log-mach=%s' % LOCAL_LOG_FILEPATH,
+            '--log-wptreport=%s' % LOCAL_REPORT_FILEPATH,
+        ]
+        if args['run_path']:
```

Comment:
Ditto. What's this do? Why the `3` above and the `5` here?

Reply:
`wpt run` expects the run path as an ordered argument. It's 3 for Sauce runs because we invoke `['wpt', 'run', $platform-id ... ]` and 5 for Chrome/FF runs because we invoke `['xvfb-run', '--auto-servernum', 'wpt', 'run', $platform-id ... ]`